hail/hail/ir-gen/src/Main.scala

```diff
@@ -0,0 +1,1108 @@
+    override def generateDeclaration: String =
+      elts.map(_.generateDeclaration).mkString("(", ", ", ")")
+
+    override def constraints: Seq[String => String] =
+      elts.zipWithIndex.flatMap { case (elt, i) =>
+        elt.constraints.map(c => (t: String) => c(s"$t._${i + 1}"))
+      }
+
+    override def nChildren(self: String): NChildren =
+      elts.zipWithIndex
+        .map { case (elt, i) => elt.nChildren(s"$self._${i + 1}") }
+        .foldLeft(NChildren())(_ + _)
+
+    override def childrenSeq(self: String): ChildrenSeq =
+      elts
+        .zipWithIndex.map { case (elt, i) => elt.childrenSeq(s"$self._${i + 1}") }
+        .foldLeft[ChildrenSeq](ChildrenSeq.empty)(_ ++ _)
+
+    override def copyWithNewChildren(self: String, newChildren: ChildrenSeqSlice): String = {
+      val offsets = elts.zipWithIndex.scanLeft(NChildren()) { case (acc, (elt, i)) =>
+        acc + elt.nChildren(s"$self._${i + 1}")
+      }
+      elts.zipWithIndex.zip(offsets).map { case ((elt, i), n) =>
+        elt.copyWithNewChildren(
+          s"$self._${i + 1}",
+          newChildren.slice(n, elt.nChildren(s"$self._${i + 1}")),
+        )
+      }.mkString("(", ", ", ")")
+    }
+  }
+
+  case class IR(
+    name: String,
+    attsAndChildren: Seq[NamedAttOrChildPack],
+    traits: Seq[Trait] = Seq.empty,
+    constraints: Seq[String] = Seq.empty,
+    extraMethods: Seq[String] = Seq.empty,
+    staticMethods: Seq[String] = Seq.empty,
+    docstring: String = "",
+    hasCompanionExtension: Boolean = false,
+  ) extends IR_Interface {
+    def withTraits(newTraits: Trait*): IR = copy(traits = traits ++ newTraits)
+    def withMethod(methodDef: String): IR = copy(extraMethods = extraMethods :+ methodDef)
+    def typed(typ: String): IR = withTraits(TypedIR(typ))
+    def withConstraint(c: String): IR = copy(constraints = constraints :+ c)
+
+    def withCompanionExtension: IR = copy(hasCompanionExtension = true)
+
+    def withClassExtension: IR = withTraits(Trait(s"${name}Ext"))
+
+    def withDocstring(docstring: String): IR = copy(docstring = docstring)
+
+    private def nChildren: NChildren = attsAndChildren.foldLeft(NChildren())(_ + _.nChildren)
+
+    private def children: String =
+      attsAndChildren.foldLeft[ChildrenSeq](ChildrenSeq.empty)(_ ++ _.childrenSeq).toString
+
+    private def childrenOffsets: Seq[NChildren] =
+      attsAndChildren.scanLeft(NChildren())(_ + _.nChildren)
+
+    private def copyMethod: String = {
+      val decl = s"override def copyWithNewChildren(newChildren: IndexedSeq[BaseIR]): $name = "
+      val assertion = s"assert(newChildren.length == $nChildren)"
+      val body = name + attsAndChildren.zipWithIndex.map { case (x, i) =>
+        try
+          x.pack.copyWithNewChildren(
+            x.name,
+            ChildrenSeqSlice(
+              ChildrenSeq.Dynamic("newChildren"),
+              nChildren,
+              childrenOffsets(i),
```

Comment:
This looks like it's calling `childrenOffsets` for every `attrsAndChildren`. Is there a reason for this? Would something like the following work?
```scala
(attsAndChildren, childrenOffsets).zipped.map { (attOrChild, offset) => .. }
```